File: apps/frontend/graphql/utils/list.ts
export interface OrderBy {
	name: string
	desc?: boolean
}

export interface ListQueryParams {
	page?: number
	pageSize?: number
	search?: string
	orderBy?: OrderBy[]
}

export interface ListResponse<T> {
	count?: number
	page?: number
	pageSize?: number
	search?: string
	data?: T[]
}

export type ListQueryResult<Data = any> = [
    (params: ListQueryParams) => void,
    {
        loading: boolean;
        data?: Data;
        error?: any;
        itemNotFound?: boolean;
        called: boolean;
    }
];

File: pkg/models/list.go
package models

const (
	DefaultPageSize = 20
	DefaultPage     = 1
)

type ListRequestParams[T any] struct {
	Filter     T
	Pagination Pagination
	OrderBy    []OrderBy
}

type ListFilter struct {
	Search *string
}

type Pagination struct {
	PageSize *uint
	Page     *uint
}

type OrderBy struct {
	Name string
	Desc *bool
}

type ListResponse[T any] struct {
	Count    int64 `json:"count"`
	PageSize *uint `json:"page_size"`
	Page     *uint `json:"page"`
	Data     []T   `json:"data"`
}
